### PR TITLE
Default DynamicCodeSupport to true for CoreCLR

### LINKED ISSUE #80398: Default IsDynamicCodeSupported feature switch to true for CoreCLR
With https://github.com/dotnet/runtime/pull/80246, we added a feature switch for `RuntimeFeature.IsDynamicCodeSupported`. Previously when you published a trimmed CoreCLR app, `RuntimeFeature.IsDynamicCodeSupported` was hard-coded to `true`, so any logic switching on that property was eligible to be trimmed. For example:

https://github.com/dotnet/runtime/blob/b900a096158a1f64e8d9d47380d0101582154461/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs#L599-L601

When you publish a CoreCLR trimmed app, the `ReflectionMemberAccessor` class could be trimmed. Now that we have a feature switch, the property is no longer a constant.

We could enable this by setting `featuredefault=true` in the [ILLink.Substitutions.xml file](https://github.com/dotnet/runtime/blob/304dfe0ba6b91046733d807755b3b90e3f134a50/src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.xml#L3), but that runs into the issue described in https://github.com/dotnet/linker/issues/3176.

This issue is to track enabling this for trimmed CoreCLR applications. It is blocked by https://github.com/dotnet/linker/issues/3176.

### PR DESCRIPTION
**Pull request by sbomer**
Fixes https://github.com/dotnet/runtime/issues/80398

### CODE REVIEW

**Review by MichalStrehovsky — APPROVED** [review]

File: src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.targets
Comment (eerhardt) [inline src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.targets:58]: What's the ordering between this and 

https://github.com/dotnet/runtime/blob/a0b8890f989d2d4b37507b99d94249506f161d6f/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets#L42

Which gets imported first? Cuz whichever one does, will win. So if the Native.targets get imported first, that would be best.

Comment (eerhardt) [reply src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.targets:58]: Looks like the native AOT targets get imported first. So that's good.

https://github.com/dotnet/sdk/blob/6e4637f352d046e599d05b9090946ff818b320f1/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets#L1331-L1332

I would hope some test would fail if these ever got switched around.

**Review by eerhardt — APPROVED** [review]

File: src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.targets
Comment (MichalStrehovsky) [reply src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.targets:58]: Things would not behave as expected if this is set to true. We error out if someone sets it to true: https://github.com/dotnet/runtime/blob/5d1981ef2d3917bc65bba9dbd36d3193f91825c5/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets#L58

File: src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj:58]: Anyone have opinions about switching crossgen2 to use the package version of ILC, same as cdacreader and ILC itself? I don't know if it's useful to use the live version and it's probably more trouble than it's worth - looks like we stabilized on using the package version instead.

(This comment is not proposing work for this PR and we'd only do this after #103375 merges.)

@dotnet/ilc-contrib @am11 for thoughts.

Comment (am11) [reply src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj:58]: Pros: live version gives (an additional) validation from CG2 (a real-world app) at the PR stage.
Cons: it slows down the development a bit; since R2R has separate code paths which are normally stabilized when next preview release is nearing. e.g. https://github.com/dotnet/runtime/pull/103537#issuecomment-2171496919

I think with ongoing code unification and converging implementations, it appears that we are leaning towards fixing issues earlier during the development phase; so using live builds is a general goodness. Also, some downstream repos like aspnetcore run into issues when a bug makes its way into the package undetected, since they have direct runtime dependency without waiting for the next SDK preview release.

Comment (eerhardt) [reply src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj:58]: My typical opinion is to use "live versions" when possible.

Comment (jkotas) [reply src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj:58]: > Anyone have opinions about switching crossgen2 to use the package version of ILC

I would be fine with that. I do not think that maintaining two different schemes is worth it, and switching everything to live build is way too complicated.

> it appears that we are leaning towards fixing issues earlier during the development phase; so using live builds is a general goodness. Also, some downstream repos like aspnetcore run into issues when a bug makes its way into the package undetected, since they have direct runtime dependency without waiting for the next SDK preview release.

This will get better with unified build codeflow in not-so-distant future. The unified build codeflow is going to be dotnet/runtime -> dotnet/dotnet and the validation of that flow is going to do the multistage build that exercises live-built toolset. If AOT compiler bug sneaks in, we will see a break in that flow.

**Pull request merged**
